### PR TITLE
add support to STTS22H temperature sensor

### PR DESCRIPTION
**Pull request by avisconti**
The STTS22H is an ultralow-power, high-accuracy, digital temperature sensor offering high performance over the entire operating temperature range. This driver is based on stmemsc HAL i/f v2.3
    
https://www.st.com/resource/en/datasheet/stts22h.pdf

The PR contains modification to sensortile_box_pro board and its standard sample.
PR has been tested on sensortile_box_pro.

### CODE REVIEW

File: drivers/sensor/stts22h/stts22h.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		ret = stts22h_set_odr_raw(dev, STTS22H_25Hz);
+		break;
+	case 3:
+		ret = stts22h_set_odr_raw(dev, STTS22H_50Hz);
+		break;
+	case 4:
+		ret = stts22h_set_odr_raw(dev, STTS22H_100Hz);
+		break;
+	case 5:
+		ret = stts22h_set_odr_raw(dev, STTS22H_200Hz);
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:103]: how about converting these to a series of

`return = stts22h_set_odr_raw(dev, STTS22H_200Hz);`

then you can drop the break, ret and even the final return (the compiler should stay happy). Makes it a bit more compact, you did it in other functions and I think it's cool.

File: drivers/sensor/stts22h/stts22h.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	return ret;
+}
+
+static int stts22h_attr_set(const struct device *dev,
+			    enum sensor_channel chan,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		LOG_WRN("invalid channel.");
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:119]: log the channel `LOG_WRN("Invalid channel: %d", chan);`

File: drivers/sensor/stts22h/stts22h.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return stts22h_odr_set(dev, val);
+	default:
+		LOG_DBG("operation not supported.");
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:127]: log `attr`

File: drivers/sensor/stts22h/stts22h.c [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	const struct stts22h_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t chip_id, odr;
+
+	if (stts22h_dev_id_get(ctx, &chip_id) < 0) {
+		LOG_DBG("Failed reading chip id");
+		return -EIO;
+	}
+
+	LOG_INF("chip id 0x%x", chip_id);
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:154]: `%02x`?

File: drivers/sensor/stts22h/stts22h_trigger.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	if (!config->int_gpio.port) {
+		return -ENOTSUP;
+	}
+
+	if (trig->chan == SENSOR_CHAN_ALL) {
+		stts22h->thsld_handler = handler;
+		stts22h->thsld_trigger = trig;
+	}
+
+	return -ENOTSUP;
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h_trigger.c:41]: something's off :-) did you mean

```c
if (trig->chan != SENSOR_CHAN_ALL) {
        return -ENOTSUP;
}

stts22h->thsld_handler = handler;
stts22h->thsld_trigger = trig;

return 0;
```
?

File: drivers/sensor/stts22h/stts22h_trigger.c [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+#if defined(CONFIG_STTS22H_TRIGGER_OWN_THREAD)
+	k_sem_init(&stts22h->gpio_sem, 0, K_SEM_MAX_LIMIT);
+
+	k_thread_create(&stts22h->thread, stts22h->thread_stack,
+			CONFIG_STTS22H_THREAD_STACK_SIZE,
+			stts22h_thread, stts22h,
+			NULL, NULL, K_PRIO_COOP(CONFIG_STTS22H_THREAD_PRIORITY),
+			0, K_NO_WAIT);
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h_trigger.c:127]: set the thread name from `dev->name`?

File: drivers/sensor/stts22h/stts22h.c (hunk 4 above)
Comment (avisconti) [reply drivers/sensor/stts22h/stts22h.c:154]: correct! :-)

File: drivers/sensor/stts22h/stts22h_trigger.c (hunk 6 above)
Comment (avisconti) [reply drivers/sensor/stts22h/stts22h_trigger.c:127]: > set the thread name from `dev->name`?

which arg is the name? I don't see it in API:
```
_tid_t k_thread_create(struct k_thread *new_thread, k_thread_stack_t *stack, size_t stack_size, k_thread_entry_t entry, void *p1, void *p2, void *p3, int prio, uint32_t options, k_timeout_t delay)
```

Comment (fabiobaltieri) [reply drivers/sensor/stts22h/stts22h_trigger.c:127]: it's a separate call, `k_thread_name_set`

File: drivers/sensor/stts22h/stts22h.c [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+static int stts22h_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct stts22h_data *data = dev->data;
+	const struct stts22h_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int16_t raw_temp;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:40]: I think this is only going to compile in if asserts are enabled, you really want it to return error all the time, just

```c
if (chan != SENSOR_CHAN_ALL) {
        return -ENOTSUP;
}
```

File: drivers/sensor/stts22h/stts22h.c [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+			       enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct stts22h_data *data = dev->data;
+
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+		stts22h_temp_convert(val, data->sample_temp);
+	} else {
+		return -ENOTSUP;
+	}
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:68]:
```c
if (chan != SENSOR_CHAN_AMBIENT_TEMP) {
        return -ENOTSUP;
}

stts22h_temp_convert(val, data->sample_temp);
```

File: drivers/sensor/stts22h/stts22h_trigger.c (hunk 6 above)
Comment (fabiobaltieri) [reply drivers/sensor/stts22h/stts22h_trigger.c:131]: ^^ in case you missed it

File: drivers/sensor/stts22h/stts22h.c (hunk 8 above)
Comment (avisconti) [reply drivers/sensor/stts22h/stts22h.c:68]: yeah, as simple as this. Right.

File: drivers/sensor/stts22h/stts22h_trigger.c (hunk 6 above)
Comment (avisconti) [reply drivers/sensor/stts22h/stts22h_trigger.c:131]: Actually I did. Thanks for reminder... :-)

File: drivers/sensor/stts22h/stts22h.c (hunk 7 above)
Comment (avisconti) [reply drivers/sensor/stts22h/stts22h.c:40]: OK, even if it seems that calling ASSERT is a pretty common way for a sensor fetch.

**Comment by avisconti** [comment]
separate push to fix the conflicts

File: drivers/sensor/stts22h/stts22h_trigger.c [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	k_thread_name_set(&stts22h->thread, "stts22h");
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h_trigger.c:132]: `	k_thread_name_set(&stts22h->thread, dev->name);`

that way you can have multiple instances with multiple triggers and they get unique names

File: drivers/sensor/stts22h/stts22h.c [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	case 2:
+		return stts22h_set_odr_raw(dev, STTS22H_25Hz);
+	case 3:
+		return stts22h_set_odr_raw(dev, STTS22H_50Hz);
+	case 4:
+		return stts22h_set_odr_raw(dev, STTS22H_100Hz);
+	case 5:
+		return stts22h_set_odr_raw(dev, STTS22H_200Hz);
+	default:
+		LOG_ERR("bad frequency");
```
Comment (fabiobaltieri) [inline drivers/sensor/stts22h/stts22h.c:104]: maybe `LOG_ERR("bad frequency: %d", val->val1);`

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
> separate push to fix the conflicts. @fabiobaltieri tell me if you agree with this method

Yeah that's ideal, we even half discussed the idea of asking contributors to do it in general some time ago but that's a bit too cumbersome for some. If you don't mind that's great, some reviewers use the incremental diff, some don't but it's good to have it.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@erwango or any of the other ST folks that could give a second review here?

**Review by MaureenHelm — DISMISSED** [review]

File: samples/boards/sensortile_box_pro/sensors-on-board/src/main.c [hunk 11]
```diff
@@ -201,6 +211,29 @@ static void lis2du12_config(const struct device *lis2du12)
...
+			    SENSOR_ATTR_SAMPLING_FREQUENCY, &odr_attr) < 0) {
+		printk("Cannot set sampling frequency for STTS22H\n");
+		return;
+	}
+
+#ifdef CONFIG_STTS22H_TRIGGER
+	struct sensor_trigger trig;
+
+	trig.type = SENSOR_TRIG_DATA_READY;
+	trig.chan = SENSOR_CHAN_ACCEL_XYZ;
```
Comment (kartben) [inline samples/boards/sensortile_box_pro/sensors-on-board/src/main.c:232]: this looks like a copy-paste error?

**Review by kartben — CHANGES_REQUESTED** [review]

File: samples/boards/sensortile_box_pro/sensors-on-board/src/main.c (hunk 11 above)
Comment (kartben) [reply samples/boards/sensortile_box_pro/sensors-on-board/src/main.c]: > sure! Thanks for getting this typo!

ChatGPT did :)

**Review by kartben — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**